core/src/air/builder.rs

```diff
@@ -368,79 +350,121 @@ pub trait MemoryAirBuilder: BaseAirBuilder {
+        let prev_values = once(prev_shard)
+            .chain(once(prev_clk))
+            .chain(once(addr.clone()))
             .chain(memory_access.prev_value().clone().map(Into::into))
             .collect();
-        let current_values = once(current_shard_expr)
-            .chain(once(current_clk_expr))
-            .chain(once(addr_expr.clone()))
+        let current_values = once(shard)
+            .chain(once(clk))
+            .chain(once(addr.clone()))
             .chain(memory_access.value().clone().map(Into::into))
             .collect();
 
         // The previous values get sent with multiplicity * 1, for "read".
         self.send(AirInteraction::new(
             prev_values,
-            verify_memory_access_expr.clone(),
+            do_check.clone(),
             InteractionKind::Memory,
         ));
 
         // The current values get "received", i.e. multiplicity = -1
         self.receive(AirInteraction::new(
             current_values,
-            verify_memory_access_expr.clone(),
+            do_check.clone(),
             InteractionKind::Memory,
         ));
     }
 
+    /// Verifies the memory access timestamp.
+    ///
+    /// This method verifies that the diff between the current and previous memory access ts is
+    /// valid.  Specifically it will ensure the following.
+    /// If the previous memory access is within the same shard, then the current_mem_clk_ts -
+    /// prev_mem_clk_ts is within [1, MAX_SHARD_CLK].
+    /// If the previous memory access is in a different shard, then the current_mem_shard_ts -
+    /// prev_mem_shard_ts is within [1, MAX_SHARD_CLK].
+    fn verify_mem_access_ts<Eb, Everify, EShard, EClk>(
+        &mut self,
+        mem_access: &MemoryAccessCols<Eb>,
+        do_check: Everify,
+        shard: EShard,
+        clk: EClk,
+    ) where
+        Eb: Into<Self::Expr> + Clone,
+        Everify: Into<Self::Expr>,
+        EShard: Into<Self::Expr>,
+        EClk: Into<Self::Expr>,
+    {
+        let do_check: Self::Expr = do_check.into();
+        let compare_clk: Self::Expr = mem_access.compare_clk.clone().into();
+        let shard: Self::Expr = shard.into();
+        let prev_shard: Self::Expr = mem_access.prev_shard.clone().into();
+
+        // First verify that compare_clk's value is correct.
+        self.when(do_check.clone()).assert_bool(compare_clk.clone());
+        self.when(do_check.clone())
+            .when(compare_clk.clone())
+            .assert_eq(shard.clone(), prev_shard);
+
+        // Get the comparison timestamp values for the current and previous memory access.
+        let prev_comp_value = self.if_else(
+            mem_access.compare_clk.clone(),
+            mem_access.prev_clk.clone(),
+            mem_access.prev_shard.clone(),
+        );
+
+        let current_comp_val = self.if_else(compare_clk.clone(), clk.into(), shard);
+
+        // Assert `current_comp_val > prev_comp_val`. We check this by asserting that
```

Comment:
nit: long comments should be written in paragraph form, similar to function declarations.

```
// Assert `current_comp_val > prev_comp_val`. We check this by asserting that
// `0 <= current_comp_val-prev_comp_val-1 < 2^24`.
//
// The equivalence of these statements comes from the fact that if `current_comp_val <= prev_comp_val`,
// then `current_comp_val-prev_comp_val-1 < 0` and will underflow in the prime field,
// resulting in a value that is `>= 2^24` as long as both `current_comp_val, prev_comp_val` are
// range-checked to be `<2^24` and as long as we're working in a field larger than `2 * 2^24`
// (which is true of the BabyBear and Mersenne31 prime).